### PR TITLE
[14.0][ADD] accout move service (pdf + facturae)

### PR DESCRIPTION
**Pull request by FrankC013**
[ADD] account_move_service_report

[IMP] sale_order_service: change 'service date' from type Date to Datetime & creation of new fields for ambugest

[IMP] connector_ambugest: new fields to import

[FIX] account_invoice_report_service: display only the date of the service date

[FIX] l10n_es_facturae_service: correction of the tests after converting the service date to datetime

### CODE REVIEW

**Review by eantones — COMMENTED** [review]
It's ok for now. We'll make changes in the next functionality

File: account_move_service_report/README.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
+===========================
```
Comment (eantones) [inline account_move_service_report/README.rst]: Rename the module to `account_move_service_report_xlsx`

File: sale_order_service/models/sale_order.py [hunk 2]
```diff
@@ -18,8 +18,15 @@ class SaleOrder(models.Model):
...
     destination = fields.Char(string="Destination")
     round_trip_code = fields.Integer(string="Round trip code")
     return_service = fields.Boolean(string="Return service")
+    service_key = fields.Char(string="Service key")
+    service_transfer_reason = fields.Char(string="Service transfer reason")
+    service_invoice_number = fields.Char(string="Service invoice number")
+    service_invoice_creation_date = fields.Date(string="Service invoice creation date")
+    service_code = fields.Char(string="Service code")
+    service_name = fields.Char(string="Service name")
+    service_type = fields.Char(string="Service name")
```
Comment (eantones) [inline sale_order_service/models/sale_order.py]:
```suggestion
    service_insurer_code = fields.Char(string="Service insurer code")
    service_insurer_name = fields.Char(string="Service insurer name")
```
What is this "service_type", it seems "Tipo de servicio" which is the result of the mapping between "Clave" and "Motivo_de_Traslado",this is not stored, it should computed on-the-fly on xlsx generation.

File: sale_order_service/views/sale_order_views.xml [hunk 3]
```diff
@@ -63,6 +63,34 @@
...
+                                attrs="{'readonly': [('state', '!=', 'draft')]}"
+                            />
+                            <field
+                                name="service_name"
+                                attrs="{'readonly': [('state', '!=', 'draft')]}"
+                            />
+                            <field
+                                name="service_type"
+                                attrs="{'readonly': [('state', '!=', 'draft')]}"
+                            />
```
Comment (eantones) [inline sale_order_service/views/sale_order_views.xml]:
```suggestion
                            <field
                                name="service_insurer_code"
                                attrs="{'readonly': [('state', '!=', 'draft')]}"
                            />
                            <field
                                name="service_insurer_name"
                                attrs="{'readonly': [('state', '!=', 'draft')]}"
                            />
```

File: connector_ambugest/models/sale_order/adapter.py [hunk 4]
```diff
@@ -11,14 +11,17 @@ class SaleOrderAdapter(Component):
...
     _sql = """select c.EMPRESA,  c."Codi UP" as "CodiUP",
-                     s.Fecha_Servicio, s.Codigo_Servicio, s.Servicio_Dia, s.Servicio_Ano,
-                     s.Fecha_Modifica,
-
+                     s.Fecha_Servicio, s.Hora_Servicio, s.Codigo_Servicio,
+                     s.Servicio_Dia, s.Servicio_Ano, s.Fecha_Modifica,
                      s.Num_Contrato, s.Nombre_Asegurado, s.DNI_Asegurado,
-                         s.Num_Asegurado,s.Referencia_autorizacion,
+                     s.Num_Asegurado, a.Codigo_Aseguradora, a.Nombre_Aseguradora,
+                     s.Referencia_autorizacion,
```
Comment (eantones) [inline connector_ambugest/models/sale_order/adapter.py]:
```suggestion
                     s.Referencia_autorizacion,
                     s.Clave, s.Motivo_Traslado,
```

File: connector_ambugest/models/sale_order/mapper.py [hunk 5]
```diff
@@ -100,11 +100,31 @@ class SaleOrderImportMapper(Component):
...
         ("Servicio_Ano", "service_number"),
-        ("Fecha_Servicio", "service_date"),
         (nullif("Origen"), "origin"),
         (nullif("Destino"), "destination"),
+        ("Clave", "service_key"),
+        ("Motivo_Traslado", "service_transfer_reason"),
+        ("Odoo_Numero_Factura", "service_invoice_number"),
+        ("Odoo_Fecha_Generada_Factura", "service_invoice_creation_date"),
+        (nullif("Codigo_Aseguradora"), "service_code"),
+        (nullif("Nombre_Aseguradora"), "service_name"),
```
Comment (eantones) [inline connector_ambugest/models/sale_order/mapper.py]:
```suggestion
        (nullif("Codigo_Aseguradora"), "service_insurer_code"),
        (nullif("Nombre_Aseguradora"), "service_insurer_name"),
```

File: l10n_es_facturae_service/models/sale_order.py [hunk 6]
```diff
@@ -16,5 +16,7 @@ def _prepare_invoice_line(self, **optional_values):
             facturae_file_reference=order.insured_ident_cardnum,
             facturae_receiver_transaction_reference=order.policy_number,
             facturae_receiver_contract_reference=order.auth_number,
-            facturae_receiver_transaction_date=order.service_date,
+            facturae_receiver_transaction_date=fields.Date.context_today(
+                order, order.service_date
+            ),
```
Comment (eantones) [inline l10n_es_facturae_service/models/sale_order.py]: Is this getting the TZ from the user who created the order? It seems it's the current user. Please check it out

File: l10n_es_facturae_service/tests/test_facturae_service.py [hunk 7]
```diff
@@ -44,6 +45,6 @@ def test_facturae_data(self):
             invoice.invoice_line_ids.mapped("facturae_receiver_contract_reference")[0],
         )
         self.assertEqual(
-            self.sale_order.service_date,
+            fields.Date.context_today(self.sale_order, self.sale_order.service_date),
```
Comment (eantones) [inline l10n_es_facturae_service/tests/test_facturae_service.py]: Same as before

File: account_invoice_report_service/models/sale_order.py [hunk 8]
```diff
@@ -19,3 +19,9 @@ def _compute_date_order_tz(self):
             datetime_utc = pytz.utc.localize(order.date_order)
             datetime_local = datetime_utc.astimezone(user_tz)
             order.date_order_tz = datetime_local.date()
+
+    service_only_date = fields.Date(compute="_compute_service_only_date")
```
Comment (eantones) [inline account_invoice_report_service/models/sale_order.py]:
```suggestion
    service_date_tz = fields.Date(compute="_compute_service_date_tz")
```

File: account_invoice_report_service/models/sale_order.py [hunk 9]
```diff
@@ -19,3 +19,9 @@ def _compute_date_order_tz(self):
...
+
+    def _compute_service_only_date(self):
+        for rec in self:
+            rec.service_only_date = fields.Date.context_today(rec, rec.service_date)
```
Comment (eantones) [inline account_invoice_report_service/models/sale_order.py]:
```suggestion
    def _compute_service_date_tz(self):
        for rec in self:
            rec.service_date_tz = fields.Date.context_today(rec, rec.service_date)
```
Here we get the TZ from the current user, it's ok as we discussed

File: account_invoice_report_service/views/report_invoice_service.xml [hunk 10]
```diff
@@ -356,7 +356,7 @@
                                                                 </td>
                                                                 <td width="10%">
                                                                     <span
-                                                                        t-field="order.service_date"
+                                                                        t-field="order.service_only_date"
```
Comment (eantones) [inline account_invoice_report_service/views/report_invoice_service.xml]:
```suggestion
                                                                        t-field="order.service_date_tz"
```

File: account_move_service_report/models/account_move.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    def print_report_invoice_service(self):
+        self.ensure_one()
+        self.check_consistency_service_report_values()
+        report_name = "account_move_service_report.report_sale_xlsx"
+        iar = self.env["ir.actions.report"]
+        report = iar._get_report_from_name(report_name)._render_xlsx(self.ids, data={})
+        move_attachments = (
+            self.env["ir.attachment"]
+            .search([("res_id", "in", self.ids), ("res_model", "=", self._name)])
+            .filtered(lambda x: x.name == "%s-Sale-Service.xlsx" % self.name)
```
Comment (eantones) [inline account_move_service_report/models/account_move.py]: "%s-Sale-Service.xlsx" Hard coded ??

File: account_move_service_report/models/account_move.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        )
+        if move_attachments:
+            move_attachments.write({"datas": base64.b64encode(report[0])})
+        else:
+            self.env["ir.attachment"].create(
+                {
+                    "name": "%s-Sale-Service.xlsx" % self.name,
```
Comment (eantones) [inline account_move_service_report/models/account_move.py]: "%s-Sale-Service.xlsx" Hard coded ??

File: account_move_service_report/report/report_sale_xlsx.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            _("ADDITIONAL (concept)"): [lambda x: x.get_service_additional_concept()],
+            _("ADDITIONAL (€)"): [lambda x: x.get_service_type_subtotal("additional")],
+            _("WAITING TIME (€)"): [lambda x: x.get_service_type_subtotal("wait")],
+            _("Total by ID (€)"): [lambda x: x.get_service_total_by("auth_number")],
+            _("Total Insurance (€)"): [
+                lambda x: x.get_service_total_by("service_code")
+            ],
+            _("Patient (name and surname)"): ["insured_name"],
+            _("Invoice Date"): ["service_invoice_creation_date"],
+            _("Invoice Number"): ["service_invoice_number"],
```
Comment (eantones) [inline account_move_service_report/report/report_sale_xlsx.py]: Get these fields from `account_move` model directly, these

```suggestion
            _("Invoice Date"): ["date"],
            _("Invoice Number"): ["name"],
```

File: account_move_service_report/report/report_sale_xlsx.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+    def report_values(self):
+        return {
+            _("Provider name"): ["company_id", "name"],
+            _("ID Number (8 digits)"): ["auth_number"],
+            _("Service Type"): [lambda x: x.get_service_typology_name()],
+            _("Service Date"): ["service_date", lambda x: x.date()],
+            _("Service Time"): ["service_date", lambda x: x.time()],
+            _("Insurance Name"): ["service_name"],
+            _("Insurance Code"): ["service_code"],
```
Comment (eantones) [inline account_move_service_report/report/report_sale_xlsx.py]:
```suggestion
            _("Insurance Name"): ["service_insurer_name"],
            _("Insurance Code"): ["service_insurer_code"],
```

File: account_move_service_report/report/report_sale_xlsx.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            _("€/KM"): [lambda x: x.get_service_type_weighted_average_price("km")],
+            _("Total KM (€)"): [lambda x: x.get_service_type_amount_total("km")],
+            _("DEPARTURE (€)"): [lambda x: x.get_service_return_price_subtotal(False)],
+            _("RETURN (€)"): [lambda x: x.get_service_return_price_subtotal(True)],
+            _("ADDITIONAL (concept)"): [lambda x: x.get_service_additional_concept()],
+            _("ADDITIONAL (€)"): [lambda x: x.get_service_type_subtotal("additional")],
+            _("WAITING TIME (€)"): [lambda x: x.get_service_type_subtotal("wait")],
+            _("Total by ID (€)"): [lambda x: x.get_service_total_by("auth_number")],
+            _("Total Insurance (€)"): [
+                lambda x: x.get_service_total_by("service_code")
```
Comment (eantones) [inline account_move_service_report/report/report_sale_xlsx.py]:
```suggestion
                lambda x: x.get_service_total_by("service_insurer_code")
```

**Review by eantones — CHANGES_REQUESTED** [review]

File: account_move_service_report/report/report_sale_xlsx.py (hunk 13 above)
Comment (FrankC013) [reply account_move_service_report/report/report_sale_xlsx.py]: It's invoice_date instead of date, right?
image

**Review by eantones — APPROVED** [review]

**Review by eantones — COMMENTED** [review]
@FrankC013 I did the merge of Ambugest, and rebase and push force your PR

**Review by eantones — APPROVED** [review]

File: account_move_service_report_xlsx/models/res_partner_service_report_config_type.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartnerServiceReportConfigType(models.Model):
+    _name = "res.partner.service.report.config.type"
+    _description = "Res Partner Service Report Config Type"
+
+    config_id = fields.Many2one(comodel_name="res.partner.service.report.config")
```
Comment (eantones) [inline account_move_service_report_xlsx/models/res_partner_service_report_config_type.py]:
```suggestion
    config_id = fields.Many2one(comodel_name="res.partner.service.report.config", ondelete='cascade', required=True)
```

File: account_move_service_report_xlsx/models/res_partner_service_report_config_tipology.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartnerServiceReportConfigTypology(models.Model):
+    _name = "res.partner.service.report.config.typology"
+    _description = "Res Partner Service Report Config Typology"
+
+    config_id = fields.Many2one(comodel_name="res.partner.service.report.config")
```
Comment (eantones) [inline account_move_service_report_xlsx/models/res_partner_service_report_config_tipology.py]:
```suggestion
    config_id = fields.Many2one(comodel_name="res.partner.service.report.config", ondelete='cascade', reuired=True)
```

File: account_move_service_report_xlsx/models/res_partner_service_report_config_tipology.py [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    name = fields.Char()
+    key = fields.Char()
+    transfer_reason = fields.Char()
```
Comment (eantones) [inline account_move_service_report_xlsx/models/res_partner_service_report_config_tipology.py]:
```suggestion
    name = fields.Char(required=True)
    key = fields.Char(required=True)
    transfer_reason = fields.Char(required=True)
```

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]

**Comment by eantones** [comment]
@FrankC013 please squash. Thanks! :-)

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]

File: account_move_service/views/report_invoice_document.xml [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+        priority="100"
+    >
+        <xpath expr="//span[@t-field='line.quantity']" position="replace">
+            <t t-if="line.intermediary_service_line">
+                <span t-esc="1" />
+            </t>
+            <t t-else="">
+                <span t-field="line.quantity" />
+            </t>
+        </xpath>
```
Comment (eantones) [inline account_move_service/views/report_invoice_document.xml]: Did you check the negative case?

File: account_move_service/views/report_invoice_document.xml [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+        <xpath expr="//span[@t-field='line.price_unit']" position="replace">
+            <t t-if="line.intermediary_service_line">
+                <span t-field="line.price_subtotal" />
+            </t>
+            <t t-else="">
+                <span t-field="line.price_unit" />
```
Comment (eantones) [inline account_move_service/views/report_invoice_document.xml:32]: Did you check the negative case?

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]

**Pull request merged**